File: framework/languages/node/e2e/RuntimeMonitoring/Client/Scenarios/mon-a1-socket-events-scenario.ts
// MON-A1: Host와 RouteMesh status를 각각 읽는다 시나리오를 검증한다.
import type { EvidenceWaitReq, ProfileReq, ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { type ManagedProcess, startServiceB, waitForPortState } from '../Support/managed-service';
import {
  hostStatusesFromEvidence,
  readHostStatus,
  readRouteStatus,
  routeStatusesFromEvidence,
  waitForRouteStatus,
  delay
} from '../Support/public-status';
import { ensure } from '../Support/scenario-assert';

export async function runMonA1(options: ClientOptions): Promise<ManagedProcess> {
  const hostBefore = await readHostStatus(options.serviceUrl);
  const routeBefore = await readRouteStatus(options.serviceUrl);
  ensure(hostBefore.isReady && hostBefore.acceptingWork, 'MON-A1 Host was not serving before the peer transition.');

  await postJson<object>(options.serviceBUrl, '/shutdown', {});
  await waitForPortState(options.serviceBUrl, false, 'MON-A1 expected the initial svc-b process to stop.');
  const routeWithoutPeer = await waitForRouteStatus(
    options.serviceUrl,
    (status) => !status.peers.some((peer) => peer.nodeRid === 'svc-b'),
    'MON-A1 RouteMesh status retained the stopped peer.'
  );

  const restarted = startServiceB(options, 'svc-b-mon-a1');
  try {
    await waitForPortState(options.serviceBUrl, true, 'MON-A1 expected svc-b to start.');
    const routeAfter = await waitForRouteStatus(
      options.serviceUrl,
      (status) => status.peers.some((peer) => peer.nodeRid === 'svc-b' && peer.state === 1)
        && status.channels.some((channel) => channel.channelName === 'monitor.profile' && channel.isReady),
      'MON-A1 RouteMesh status did not restore the peer and Channel readiness.'
    );
    ensure(
      BigInt(routeAfter.sequence) > BigInt(routeBefore.sequence),
      'MON-A1 RouteMesh sequence did not increase in its own source.'
    );
    ensure(
      JSON.stringify(routeBefore) !== JSON.stringify(routeAfter),
      'MON-A1 previously returned RouteMesh status was mutated by a later transition.'
    );

    const hostAfter = await readHostStatus(options.serviceUrl);
    ensure(hostAfter.isReady && hostAfter.acceptingWork, 'MON-A1 Host stopped accepting work during peer recovery.');
    ensure(
      BigInt(hostAfter.sequence) >= BigInt(hostBefore.sequence),
      'MON-A1 Host sequence was compared outside its source or moved backwards.'
    );

    const request: ProfileReq = { value: 'monitor', marker: 'mon-a1-request' };
    const reply = await postJson<ProfileRes>(options.triggerUrl, '/profile/request/service-b', request);
    ensure(reply.value === 'profile:monitor' && reply.providerRid === 'svc-b', 'MON-A1 request did not reach the expected provider.');
    await postJson<string[]>(options.serviceBUrl, '/evidence/wait', {
      containsAll: ['profile-request|rid=svc-b|marker=mon-a1-request|value=monitor'],
      containsAnyGroups: [],
      timeoutMilliseconds: 10000
    } satisfies EvidenceWaitReq);

    await waitForObservedRoute(options.serviceUrl, (status) => BigInt(status.sequence) >= BigInt(routeAfter.sequence));
    await waitForObservedHost(options.serviceUrl, (status) => BigInt(status.sequence) >= BigInt(hostAfter.sequence));
    console.log('scenario MON-A1 passed');
    return restarted;
  } catch (error) {
    await restarted.stop();
    throw error;
  }
}

async function waitForObservedRoute(
  url: string,
  predicate: (status: ReturnType<typeof routeStatusesFromEvidence>[number]) => boolean
): Promise<void> {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const statuses = routeStatusesFromEvidence(await getJson<readonly string[]>(url, '/evidence'));
    if (statuses.some(predicate)) return;
    await delay();
  }
  throw new Error('MON-A1 RouteMesh observer did not provide the current public status.');
}

async function waitForObservedHost(
  url: string,
  predicate: (status: ReturnType<typeof hostStatusesFromEvidence>[number]) => boolean
): Promise<void> {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const statuses = hostStatusesFromEvidence(await getJson<readonly string[]>(url, '/evidence'));
    if (statuses.some(predicate)) return;
    await delay();
  }
  throw new Error('MON-A1 Host observer did not provide the current public status.');
}

File: framework/languages/dotnet/e2e/RuntimeMonitoring/Client/Scenarios/MonA1SocketEventsScenario.cs
// Verifies Config 7 MON-A1 with two immutable public RouteMesh statuses.
using System.Diagnostics;
using RuntimeMonitoring.Client.Support;
using RuntimeMonitoring.Shared;
using Zlink.HttpClient;

namespace RuntimeMonitoring.Client.Scenarios;

internal static class MonA1SocketEventsScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        using var observer = ZLinkHttpClient.Create(options.ServiceUrl)
            .Timeout(TimeSpan.FromSeconds(35))
            .Build();

        var baseline = await SnapshotAsync(observer);
        AssertBaseline(baseline);
        var evidenceBaseline =
            (await observer.Get("/evidence").Async<string[]>()).Body.Length;

        await using var serviceB = await EphemeralService.StartAsync(options, "svc-b");
        var ready = await WaitForPeerAsync(observer, "svc-b");
        var peer = ready.Peers.Single(candidate =>
            candidate.Rid.StartsWith("svc-b-", StringComparison.Ordinal)
            && candidate.State == "Ready");

        ZlinkStreamAssert.Ensure(
            ready.Sequence > baseline.Sequence
            && baseline.Peers.Length == 0
            && peer.UnavailableReason is null,
            "MON-A1 immutable status or ready peer state was incomplete.");

        var channel = ready.Channels.Single(candidate =>
            candidate.ChannelName == RuntimeMonitoringNames.Channel);
        ZlinkStreamAssert.Ensure(
            channel.IsReady && channel.ReadyTargetCount == 2,
            "MON-A1 channel status did not include both ready targets.");

        var evidence = (await observer.Post("/evidence/wait")
            .Body(new EvidenceWaitReq(
                [
                    "identifier=zlink.runtime.mesh_node.peer_changed",
                    "routing=svc-b",
                    "kind=ConnectionReady"
                ],
                [],
                TimeoutMilliseconds: 3000,
                AfterIndex: evidenceBaseline))
            .Async<string[]>()).Body;
        ZlinkStreamAssert.Ensure(
            evidence.Any(line =>
                line.Contains("identifier=zlink.runtime.mesh_node.peer_changed",
                    StringComparison.Ordinal)
                && line.Contains("routing=svc-b", StringComparison.Ordinal)
                && line.Contains("sequence=", StringComparison.Ordinal)),
            "MON-A1 typed peer event was not observed.");

        Console.WriteLine("scenario MON-A1 passed");
    }

    private static void AssertBaseline(MeshRuntimeSnapshotRes status)
    {
        var channel = status.Channels.Single(candidate =>
            candidate.ChannelName == RuntimeMonitoringNames.Channel);
        ZlinkStreamAssert.Ensure(
            status.MeshName == RuntimeMonitoringNames.Channel
            && status.State == "Ready"
            && status.IsReady
            && status.ReadyPeerCount == 0
            && status.Sequence > 0
            && status.ObservedAt != default
            && status.Peers.Length == 0
            && channel.IsReady
            && channel.ReadyTargetCount == 1
            && !status.Placement.IsAvailable,
            "MON-A1 baseline RouteMesh status was incomplete.");
    }

    private static async Task<MeshRuntimeSnapshotRes> SnapshotAsync(
        ZLinkHttpClient service)
        => (await service.Get(
                $"/runtime/snapshot/{RuntimeMonitoringNames.Channel}")
            .Async<MeshRuntimeSnapshotRes>()).Body;

    private static async Task<MeshRuntimeSnapshotRes> WaitForPeerAsync(
        ZLinkHttpClient service,
        string rid)
    {
        var elapsed = Stopwatch.StartNew();
        while (true)
        {
            var status = await SnapshotAsync(service);
            if (status.Peers.Any(peer =>
                    peer.Rid.StartsWith($"{rid}-", StringComparison.Ordinal)
                    && peer.State == "Ready"))
                return status;
            if (elapsed.Elapsed >= TimeSpan.FromSeconds(3))
                throw new InvalidOperationException(
                    $"MON-A1 peer '{rid}' did not become ready.");
            await Task.Delay(TimeSpan.FromMilliseconds(50));
        }
    }
}
